src/azul/service/avro_pfb.py

```diff
@@ -0,0 +1,451 @@
+class Relation:
+    dst_id: str
+    # A more appropriate variable name would be dst_type, but we stick with name
+    # to conform to PFB spec
+    dst_name: str
+
+    @classmethod
+    def to_entity(cls, entity: 'PFBEntity'):
+        return cls(dst_id=entity.id, dst_name=entity.name)
+
+
+pfb_namespace = uuid.UUID('3372fc62-76d2-4ada-a68a-f08d4f558da5')
+
+
+@attr.s(auto_attribs=True, frozen=True, kw_only=True)
+class PFBEntity:
+    """
+    Python representation of the PFB data object. Attribute names conform to the
+    PFB spec (which simplifies serialization).
+    """
+    id: str
+    name: str
+    object: MutableJSON = attr.ib(eq=False)
+
+    @classmethod
+    def from_no_id(cls, name: str, object_: MutableJSON) -> 'PFBEntity':
+        """
+        Derive id from object in a reproducible way so that we can distinguish
+        entities by comparing their ids.
+
+        Entities need to be quickly distinguishable, but unfortunately
+        document_id wasn't unique (or present) for all entities. Thus, we use
+        the frozen state of the JSON to generate our hash and check equality.
+        """
+        entity_id = str(uuid.uuid5(pfb_namespace, repr(sort_frozen(freeze(object_)))))
+        return cls(id=entity_id, name=name, object=object_)
+
+    def to_json(self, relations: Iterable[Relation]):
+        return {
+            **attr.asdict(self),
+            "relations": [attr.asdict(relation) for relation in relations]
+        }
+
+
+class EntityAggregator:
+
+    def __init__(self):
+        self._entities_to_relations: MutableMapping[PFBEntity, MutableSet[Relation]] = defaultdict(set)
+
+    def add_doc(self, doc: JSON):
+        """
+        For an Elasticsearch document
+        """
+        contents = doc['contents']
+        # File entities are assumed to be unique
+        file_entity = PFBEntity.from_no_id(name='files', object_=one(contents['files']))
+        assert file_entity not in self._entities_to_relations
+        # File entities won't relate *to* anything, but things relate *to* them.
+        # Terra streams PFBs and thus entities must be defined before they are
+        # referenced. This is one reason that file_entity is added first.
+        self._entities_to_relations[file_entity] = set()
+        for entity_type, objects in contents.items():
+            if entity_type != 'files':
+                if len(objects) == 0:
+                    continue
+                elif len(objects) == 1:
+                    entity = PFBEntity.from_no_id(name=entity_type, object_=one(objects))
+                    self._entities_to_relations[entity].add(Relation.to_entity(file_entity))
+                else:
+                    raise ValueError(f'Elasticsearch document {doc} had unexpected {entity_type!r} entities')
+
+    def entities(self, schema: JSON) -> Iterable[PFBEntity]:
+        for entity, relations in self._entities_to_relations.items():
+            add_missing_fields(entity, schema)
```

Comment:
It's a bit odd that this class mutates the objects it contains. Filling in fields seems like an unrelated concern to what this class.